akka-projection-core/src/main/scala/akka/projection/internal/OffsetStrategy.scala

```diff
@@ -50,50 +50,82 @@ private[projection] final case class AtLeastOnce(
  * INTERNAL API
  */
 @InternalApi
-private[projection] sealed trait HandlerStrategy[Envelope] {
+private[projection] sealed trait HandlerStrategy {
+  def recreateHandlerOnNextAccess(): Unit
+
   def lifecycle: HandlerLifecycle
 
   def actorHandlerInit[T]: Option[ActorHandlerInit[T]]
 }
 
-/**
- * INTERNAL API
- */
 @InternalApi
-private[projection] final case class SingleHandlerStrategy[Envelope](handler: Handler[Envelope])
-    extends HandlerStrategy[Envelope] {
-  override def lifecycle: HandlerLifecycle = handler
+private[projection] sealed abstract class FunctionHandlerStrategy[Envelope](handlerFactory: () => Handler[Envelope])
+    extends HandlerStrategy {
+  @volatile private var _handler: Option[Handler[Envelope]] = None
+  @volatile private var _recreateHandlerOnNextAccess = true
+
+  def recreateHandlerOnNextAccess(): Unit =
+    _recreateHandlerOnNextAccess = true
 
-  override def actorHandlerInit[T]: Option[ActorHandlerInit[T]] = handler match {
+  /**
+   * Current handler instance, or lazy creation of it.
+   */
+  def handler(): Handler[Envelope] = {
+    _handler match {
+      case Some(h: Handler[Any] @unchecked) if !_recreateHandlerOnNextAccess => h
+      case _ =>
+        createHandler()
+        _recreateHandlerOnNextAccess = false
+        _handler.get
```

Comment:
It's not clear for me why all that recreate dance. 

In which case, `handler()` will be called, `_handler` will be a `Some` and `_recreateHandlerOnNextAccess` will be `false`? 

We start as (None, true), then we move to (Some, false).

If the source terminates, it becomes (Some, true). Line 75 won't be executed and we will recreate it again (eventually preserving the actor). 

The only case I see where we could call `handler()` when the state is (Some, false) is when there is a bug which then makes me feel that we are being too defensive. 
If so, I would prefer no to add it. It adds complexity and it makes it hard to understand the motivation.

Or am I missing something?

Reply:
`_recreateHandlerOnNextAccess = true` when the stream fails, so that the handler is recreated when the stream is restarted (next access).

It's lazy creation, but `handler()` may also be called several times and in second call it will be (some, false).

Reply:
I understood that the whole mechanism is to allow us to call `handler()` more the once like if it was a lazy val. My point is that I don't think it can happen unless we build an ill-behaved projection.

Reply:
it's definitely called more than once when the projection is created first time, once via `actorHandlerInit`.

Then it's also called via `lifecycle` and via `single.handler()`, so I'm pretty sure it's called several times also for restarts. Treating it as a lazy is pretty ok to reason about, although the internals here are indeed a bit ugly.

Reply:
Ok, I missed that. I thought we called it only once per start.